File: completed/0169.java
class Solution {
    public int majorityElement(int[] nums) {
        // boyer-moore majority vote algorithm
        int m = 0;
        int c = 0;
        for (int num : nums) {
            if (c == 0) {
                m = num;
                c = 1;
            } else if (m == num) {
                c++;
            } else {
                c--;
            }
        }
        return m;
    }

    public static void main(String[] args) {
        int[] nums = { 3, 2, 3 };

        Solution solution = new Solution();
        System.out.println(solution.majorityElement(nums));
    }
}

File: completed/0169.go
package main

import "fmt"

func majorityElement(nums []int) int {
    m := make(map[int]int)
    for i := range nums {
        n := nums[i]
        _, p := m[n]
        if !p {
            m[n] = 0
        }
        m[n]++
    }
    majority := nums[0]
    for k, v := range m {
        if v > m[majority] {
            majority = k
        }
    }
    return majority
}

func main() {
    nums := []int{2,2,1,1,1,2,2}
    fmt.Println(majorityElement(nums))
}
